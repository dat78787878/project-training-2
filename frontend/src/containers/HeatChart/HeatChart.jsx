import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import { getData } from '../../redux/heatChart/actions';
import Chart from 'react-apexcharts';
import Loading from '../../components/Loading/Loading';
import Error from '../../components/Error/Error';
import { Col, Row } from 'react-bootstrap';

const HeatChart = () => {
  const { heatData, isLoading, isError } = useSelector((state) => state.heatChart);
  const { fromDate, toDate } = useSelector((state) => state.dateRange);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getData([fromDate, toDate]));
  }, [fromDate, toDate]);

  const dataBarChart = heatData.map((value) => {
    return value.data.reduce((sum, currentVal) => {
      return sum + currentVal.y;
    }, 0);
  });
  const optionsBar = {
    chart: {
      offsetX: -30,
      toolbar: {
        show: false
      },
      stacked: true
    },
    plotOptions: {
      bar: {
        horizontal: true,
        borderRadius: 10,
        barHeight: '90%'
      }
    },
    colors: '#0aff01',
    yaxis: {
      labels: {
        show: false
      }
    },
    dataLabels: {
      enabled: true,
      textAnchor: 'start',
      style: {
        colors: ['#fff']
      },
      offsetX: 0,
      dropShadow: {
        enabled: true
      }
    },
    xaxis: {
      axisTicks: {
        show: false
      },
      min: 0,
      categories: [
        'Sunday',
        'Saturday',
        'Friday',
        'Thursday',
        'Wednesday',
        'Tuesday',
        'Monday'
      ].reverse(),
      labels: {
        show: true,
        hideOverlappingLabels: true,
        formatter: (value) => {
          return value.toFixed(0);
        },
        style: {
          fontSize: '14px'
        }
      }
    },

    grid: {
      show: false
    }
  };

  const seriesBar = [
    {
      data: dataBarChart.reverse()
    }
  ];

  const options = {
    chart: {
      toolbar: {
        show: false
      }
    },
    dataLabels: {
      enabled: false
    },

    colors: ['#b700ff'],
    xaxis: {
      labels: {
        formatter: (value) => {
          return value % 12;
        },
        style: {
          fontSize: '14px'
        }
      }
    },
    yaxis: {
      opposite: true,
      labels: {
        align: 'center',
        offsetX: 5,
        formatter: (value) => {
          return value.toString().substring(0, 3);
        },
        style: {
          fontSize: '14px'
        }
      }
    }
  };

  const heatChart = (
    <div className="heatChart">
      <h5 className="">Device By Hour</h5>
      <div className="heatChart-detail">
        <Chart options={options} series={heatData} type="heatmap" height={240} />
      </div>
      <div className="d-flex justify-content-between heatChart-time">
        <span>am</span>
        <span>pm</span>
      </div>
      <div className="heatChart-line"></div>
      <div className="d-flex justify-content-between heatChart-number">
        <span>0</span>
        <span>5</span>
        <span>10</span>
        <span>15</span>
      </div>
    </div>
  );

  const barChart = (
    <div className="barChart">
      <Chart options={optionsBar} series={seriesBar} type="bar" height={230} />
    </div>
  );

  return (
    <div className="heatChart">
      {isLoading && <Loading />}
      {isError && <Error />}
      {!isError && !isLoading && (
        <Row>
          <Col xs={9}>{heatChart}</Col>
          <Col xs={3}>{barChart}</Col>
        </Row>
      )}
    </div>
  );
};

export default HeatChart;
