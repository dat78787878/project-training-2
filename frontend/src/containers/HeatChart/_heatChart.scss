.heatChart {
  background-color: #ffffff;
  position: relative;

  @include pc {
    height: 400px;
  }

  height: 385px;

  @include laptop {
    margin-top: 10px;
  }

  @include tablet {
    margin-top: 10px;
  }

  @include mobile {
    height: 600px;
    margin-top: 10px;
  }






  &-container {
    &-line {
      height: 8px;
      background-image: linear-gradient(to right, #ffffff, #b700ff);
      width: 91%;

      @include pc {
        width: 93%;
      }

      @include tablet {
        width: 92%;
      }

      @include mobile {
        width: 86%;
        ;
      }
    }

    &-number {
      width: 94%;
      padding-left: 13px;
    }

    &-time {
      width: 50%;
      padding-left: 13px;
    }

    .apexcharts-xaxis-texts-g text:nth-child(2n) {
      visibility: hidden;
    }
  }

}

.barChart {
  padding-top: 70px;


  @include mobile {
    padding-top: 10px;
    padding-left: 20px;
  }

  .apexcharts-xaxis-texts-g {
    text:not(:first-child):not(:last-child) {
      visibility: hidden;
    }
  }
}
