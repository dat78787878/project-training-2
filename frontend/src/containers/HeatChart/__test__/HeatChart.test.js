import React from 'react';
import { render, screen } from '@testing-library/react';
import { useSelector, useDispatch } from 'react-redux';
import HeatChart from '../HeatChart';
window.React = React;

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn()
}));
jest.mock('react-apexcharts', () => 'Chart');
describe('HeatChart Component', () => {
  let mockAppState = {};
  beforeEach(() => {
    useDispatch.mockImplementation(() => () => {});
    useSelector.mockImplementation((callback) => {
      return callback(mockAppState);
    });
  });
  afterEach(() => {
    useSelector.mockClear();
    useDispatch.mockClear();
  });
  it('check loading', () => {
    mockAppState = {
      heatChart: {
        heatData: [],
        isLoading: true,
        isError: false
      }
    };
    const { container } = render(<HeatChart />);
    expect(container).toMatchSnapshot();
    expect(container.querySelector('div').firstChild.className).toEqual('loading');
  });
  it('error task', () => {
    mockAppState = {
      heatChart: {
        heatData: [],
        isLoading: false,
        isError: true
      }
    };
    const { container } = render(<HeatChart />);
    expect(container).toMatchSnapshot();
    expect(screen.getByText('Error')).toBeTruthy();
  });
  it('success task', () => {
    mockAppState = {
      heatChart: {
        heatData: [
          {
            name: 'Monday',
            data: [
              { x: '1', y: 5 },
              { x: '2', y: 5 },
              { x: '3', y: 5 }
            ]
          },
          {
            name: 'Tuesday',
            data: [
              { x: '1', y: 6 },
              { x: '2', y: 6 },
              { x: '3', y: 6 }
            ]
          },
          {
            name: 'March',
            data: [
              { x: '1', y: 2 },
              { x: '2', y: 2 },
              { x: '3', y: 2 }
            ]
          }
        ],
        isLoading: false,
        isError: false
      }
    };
    const { container } = render(<HeatChart />);
    expect(container).toMatchSnapshot();
    expect(container.getElementsByClassName('apexcharts-canvas'));
  });
});
