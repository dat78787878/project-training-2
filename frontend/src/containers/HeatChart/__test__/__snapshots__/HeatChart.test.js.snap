// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeatChart Component check loading 1`] = `
<div>
  <div
    class="heatChart"
  >
    <div
      class="loading"
    >
      <div
        class="spinner-border text-success"
      />
    </div>
  </div>
</div>
`;

exports[`HeatChart Component error task 1`] = `
<div>
  <div
    class="heatChart"
  >
    <div
      class="error"
    >
      Error
      <img
        alt="not-found"
        height="100px"
        src="https://www.pngitem.com/pimgs/m/561-5616833_image-not-found-png-not-found-404-png.png"
        width="100px"
      />
    </div>
  </div>
</div>
`;

exports[`HeatChart Component success task 1`] = `
<div>
  <div
    class="heatChart"
  >
    <div
      class="row"
    >
      <div
        class="col-md-9 col-sm-12"
      >
        <div
          class="heatChart-container padding-title"
        >
          <h5
            class=""
          >
            Device By Hour
          </h5>
          <div
            class="heatChart-container-detail"
          >
            <chart
              height="240"
              options="[object Object]"
              series="[object Object],[object Object],[object Object]"
              type="heatmap"
            />
          </div>
          <div
            class="d-flex justify-content-between heatChart-container-time"
          >
            <span>
              am
            </span>
            <span>
              pm
            </span>
          </div>
          <div
            class="heatChart-container-line"
          />
          <div
            class="d-flex justify-content-between heatChart-container-number"
          >
            <span>
              0
            </span>
            <span>
              5
            </span>
            <span>
              10
            </span>
            <span>
              15
            </span>
          </div>
        </div>
      </div>
      <div
        class="col-md-3 col-sm-12"
      >
        <div
          class="barChart"
        >
          <chart
            height="230"
            options="[object Object]"
            series="[object Object]"
            type="bar"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
